bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -612,6 +629,51 @@ public void asyncReadUnconfirmedEntries(long firstEntry, long lastEntry, ReadCal
         asyncReadEntriesInternal(firstEntry, lastEntry, cb, ctx);
     }
 
+    /**
+     * Read a sequence of entries asynchronously.
+     *
+     * @param firstEntry
+     *          id of first entry of sequence
+     * @param lastEntry
+     *          id of last entry of sequence
+     */
+    @Override
+    public CompletableFuture<Iterable<org.apache.bookkeeper.client.api.LedgerEntry>> read(long firstEntry, long lastEntry) {
```

Comment:
ok..i briefly went through all the classes.. you are planning to make org.apache.bookkeeper.client.api as customer facing/public package/classes. But for the sake of backward compatibility you would leave existing org.apache.bookkeeper.client classes as public, right? But there is confusion, regarding the changes you are making

- do you have any plans of deprecating or making non-public, existing org.apache.bookkeeper.client classes, so that new users would be less confused and use new classes/API
- even with your new set of classes/package org.apache.bookkeeper.client.BookKeeper is the entry point for Bookkeeper API, it would be confusing for users to create instance of client.BookKeeper class and then start using bookkeeper.client.api for their operations.
- is it not necessary for new bookkeeper.client.api apis to be on par with org.apache.bookkeeper.client APIs? I dont see equivalent methods for ledgerHandle.getCustomMetadata, getNumFragments, ..

Reply:
@eolivelli  I'll wait for your comments before doing further review. In summary I want to understand relation between org.apache.bookkeeper.client package and org.apache.bookkeeper.client.api package and their corresponding classes.

I was expecting just builder interfaces for create, open and delete op as part of this change.

@jvrao FYI

Reply:
@reddycharan  thank you for your review it is very good to see that we will have a new high level API which is the result of the work of many committers !

Answers to the higher level questions:
1) Deprecation of existing API
in the BP-15 we discussed not to add @Deprecated annotations in 4.6, maybe we can in futher releases, and I think we should drop them in 5.0.
The reason is that all of the dependent projects are running with the callback based API and we will introduce a lot of warnings. We can introduce the new API softly and encourage the switch.

2) org.apache.bookkeeper.client.BookKeeper is the entry point for Bookkeeper API....
I have added a new BookKeeperBuilder into org.apache.bookkeeper.client.api, that you. The only reference to the old package is not BKException and its subclassses.
Thank you very much for pointing at this

3) is it not necessary for new bookkeeper.client.api apis to be on par with org.apache.bookkeeper.client APIs....

The new API at the moment contains only the main functions necessary to write and read for ledgers. IMHO we should add a new API for the legacy BookKeeperAdmin API:
- listing ledgers
- accessing ledgers metadata
- managing metadata in general (formatting....)

We will have to work on the PlacementPolicy API too in the future

If you think that there are some critical missing methods in the new API please write a list or submit an issue.

We should move forward on this patch as soon as possible, the next step will be to introduce the LedgerType which is critical for BP-14 (journal no-sync) and for the improvements on SF side about the no-journal ledgers.....